File: zena_mode/handlers/orchestration.py
"""
Orchestration Handler - Manages model swapping, swarm scaling, and admin operations.
"""

import logging
import threading
from zena_mode.handlers.base import BaseZenHandler

logger = logging.getLogger("ZenAI.Handler.Orchestration")


class OrchestrationHandler:
    """Namespace for orchestration-related request handling (swap, scale, etc.)."""

    @staticmethod
    def handle_post(handler: BaseZenHandler):
        """Routing for POST requests related to orchestration."""

        if handler.path == "/swap":
            from zena_mode.server import restart_with_model

            params = handler.parse_json_body()
            model_name = params.get("model")
            if not model_name:
                handler.send_json_response(400, {"error": "Missing 'model' parameter"})
                return True
            threading.Thread(target=restart_with_model, args=(model_name,), daemon=True).start()
            handler.send_json_response(200, {"status": "accepted", "model": model_name})
            return True

        if handler.path == "/swarm/scale":
            from zena_mode.server import scale_swarm

            params = handler.parse_json_body()
            count = params.get("count", 3)
            if not isinstance(count, int) or count < 1 or count > 10:
                handler.send_json_response(400, {"error": "count must be integer 1-10"})
                return True
            threading.Thread(target=scale_swarm, args=(count,), daemon=True).start()
            handler.send_json_response(200, {"status": "scaling", "target": count})
            return True

        if handler.path == "/swarm/launch":
            from zena_mode.server import launch_expert_process
            from config_system import config
            from pathlib import Path

            params = handler.parse_json_body()
            model_name = params.get("model")
            port = params.get("port")

            if not model_name or not port:
                handler.send_json_response(400, {"error": "Missing 'model' or 'port'"})
                return True

            # Resolve Model Path
            # 1. Check if absolute
            m_path = Path(model_name)
            if not m_path.is_absolute():
                # 2. Check in standard model dir
                m_path = config.MODEL_DIR / model_name

            if not m_path.exists():
                # 3. Check in central store via config (portable)
                try:
                    from config import MODEL_DIR as _cfg_model_dir
                    central = _cfg_model_dir / model_name
                except Exception:
                    central = m_path  # no-op fallback
                if central.exists():
                    m_path = central
                else:
                    handler.send_json_response(404, {"error": f"Model not found: {model_name}"})
                    return True

            # Launch in background thread to not block response?
            # launch_expert_process is fast (Popen), so it's fine.
            try:
                launch_expert_process(port=int(port), threads=2, model_path=m_path)
                handler.send_json_response(200, {"status": "launched", "port": port, "model": str(m_path)})
            except Exception as e:
                handler.send_json_response(500, {"error": str(e)})

            return True

        return False

File: rust_output/src/orchestration.rs
/// Orchestration Handler - Manages model swapping, swarm scaling, and admin operations.

use anyhow::{Result, Context};
use crate::base::{BaseZenHandler};
use std::collections::HashMap;
use std::path::PathBuf;

pub static LOGGER: std::sync::LazyLock<String /* logging::getLogger */> = std::sync::LazyLock::new(|| Default::default());

/// Namespace for orchestration-related request handling (swap, scale, etc.).
#[derive(Debug, Clone)]
pub struct OrchestrationHandler {
}

impl OrchestrationHandler {
    /// Routing for POST requests related to orchestration.
    pub fn handle_post(handler: BaseZenHandler) -> Result<()> {
        // Routing for POST requests related to orchestration.
        if handler.path == "/swap".to_string() {
            // TODO: from zena_mode.server import restart_with_model
            let mut params = handler.parse_json_body();
            let mut model_name = params.get(&"model".to_string()).cloned();
            if !model_name {
                handler.send_json_response(400, HashMap::from([("error".to_string(), "Missing 'model' parameter".to_string())]));
                true
            }
            std::thread::spawn(|| {});
            handler.send_json_response(200, HashMap::from([("status".to_string(), "accepted".to_string()), ("model".to_string(), model_name)]));
            true
        }
        if handler.path == "/swarm/scale".to_string() {
            // TODO: from zena_mode.server import scale_swarm
            let mut params = handler.parse_json_body();
            let mut count = params.get(&"count".to_string()).cloned().unwrap_or(3);
            if (!/* /* isinstance(count, int) */ */ true || count < 1 || count > 10) {
                handler.send_json_response(400, HashMap::from([("error".to_string(), "count must be integer 1-10".to_string())]));
                true
            }
            std::thread::spawn(|| {});
            handler.send_json_response(200, HashMap::from([("status".to_string(), "scaling".to_string()), ("target".to_string(), count)]));
            true
        }
        if handler.path == "/swarm/launch".to_string() {
            // TODO: from zena_mode.server import launch_expert_process
            // TODO: from config_system import config
            // TODO: from pathlib import Path
            let mut params = handler.parse_json_body();
            let mut model_name = params.get(&"model".to_string()).cloned();
            let mut port = params.get(&"port".to_string()).cloned();
            if (!model_name || !port) {
                handler.send_json_response(400, HashMap::from([("error".to_string(), "Missing 'model' or 'port'".to_string())]));
                true
            }
            let mut m_path = PathBuf::from(model_name);
            if !m_path.is_absolute() {
                let mut m_path = (config::MODEL_DIR / model_name);
            }
            if !m_path.exists() {
                // try:
                {
                    // TODO: from config import MODEL_DIR as _cfg_model_dir
                    let mut central = (_cfg_model_dir / model_name);
                }
                // except Exception as _e:
                if central.exists() {
                    let mut m_path = central;
                } else {
                    handler.send_json_response(404, HashMap::from([("error".to_string(), format!("Model not found: {}", model_name))]));
                    true
                }
            }
            // try:
            {
                launch_expert_process(/* port= */ port.to_string().parse::<i64>().unwrap_or(0), /* threads= */ 2, /* model_path= */ m_path);
                handler.send_json_response(200, HashMap::from([("status".to_string(), "launched".to_string()), ("port".to_string(), port), ("model".to_string(), m_path.to_string())]));
            }
            // except Exception as e:
            true
        }
        Ok(false)
    }
}
